File: IFATCCompanion/Views/Components.swift
import SwiftUI

/// Traffic-light status used across the dashboard.
enum StatusLevel {
    case green, amber, red, neutral

    var color: Color {
        switch self {
        case .green: return Color(red: 0.30, green: 0.82, blue: 0.45)
        case .amber: return Color(red: 0.98, green: 0.74, blue: 0.18)
        case .red: return Color(red: 0.95, green: 0.35, blue: 0.32)
        case .neutral: return Color.gray
        }
    }
}

/// A rounded dark card container.
struct Card<Content: View>: View {
    var title: String? = nil
    var systemImage: String? = nil
    @ViewBuilder var content: () -> Content

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            if let title {
                HStack(spacing: 6) {
                    if let systemImage { Image(systemName: systemImage) }
                    Text(title).font(.headline)
                }
                .foregroundStyle(.secondary)
            }
            content()
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: 16, style: .continuous)
                .fill(Color(.secondarySystemBackground))
        )
    }
}

/// A status chip with a colored dot.
struct StatusPill: View {
    var text: String
    var level: StatusLevel
    var systemImage: String? = nil

    var body: some View {
        HStack(spacing: 6) {
            Circle().fill(level.color).frame(width: 10, height: 10)
            if let systemImage { Image(systemName: systemImage).font(.caption) }
            Text(text).font(.subheadline.weight(.semibold))
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 7)
        .background(Capsule().fill(level.color.opacity(0.18)))
        .overlay(Capsule().stroke(level.color.opacity(0.5), lineWidth: 1))
    }
}

/// A label/value row for data panels.
struct DataRow: View {
    var label: String
    var value: String
    var systemImage: String? = nil

    var body: some View {
        HStack {
            HStack(spacing: 6) {
                if let systemImage { Image(systemName: systemImage).foregroundStyle(.secondary).frame(width: 22) }
                Text(label).foregroundStyle(.secondary)
            }
            Spacer()
            Text(value).font(.body.weight(.semibold)).multilineTextAlignment(.trailing)
        }
        .font(.body)
    }
}

/// Large action button suitable for one-handed use while flying.
struct ActionButton: View {
    var title: String
    var systemImage: String
    var tint: Color = .accentColor
    var action: () -> Void

    var body: some View {
        Button(action: action) {
            VStack(spacing: 6) {
                Image(systemName: systemImage).font(.title3)
                Text(title)
                    .font(.subheadline.weight(.semibold))
                    .multilineTextAlignment(.center)
                    .lineLimit(2)
                    .minimumScaleFactor(0.8)
            }
            .frame(maxWidth: .infinity, minHeight: 64)
            .padding(.vertical, 6)
        }
        .buttonStyle(.bordered)
        .tint(tint)
    }
}

/// A frequency-tune button: facility name plus the frequency it's reached on,
/// highlighted while it's the controller currently being worked. Dimmed once the
/// facility has no further call in the flight.
struct FrequencyButton: View {
    var title: String
    var systemImage: String
    var frequency: String
    var active: Bool
    var enabled: Bool
    var action: () -> Void

    private var label: some View {
        VStack(spacing: 4) {
            Image(systemName: systemImage).font(.title3)
            Text(title)
                .font(.subheadline.weight(.semibold))
                .lineLimit(1)
                .minimumScaleFactor(0.8)
            Text(frequency)
                .font(.caption2.monospacedDigit())
                .foregroundStyle(active ? Color.white.opacity(0.9) : .secondary)
                .lineLimit(1)
                .minimumScaleFactor(0.8)
        }
        .frame(maxWidth: .infinity, minHeight: 72)
        .padding(.vertical, 6)
    }

    var body: some View {
        Group {
            if active {
                Button(action: action) { label }.buttonStyle(.borderedProminent)
            } else {
                Button(action: action) { label }.buttonStyle(.bordered)
            }
        }
        .tint(.accentColor)
        .disabled(!enabled)
    }
}

extension View {
    /// Convenience to apply the standard screen background.
    func screenBackground() -> some View {
        self.background(Color(.systemBackground).ignoresSafeArea())
    }
}

File: Android/app/src/main/kotlin/com/h3consultingpartners/ifatccompanion/ui/components/Components.kt
package com.h3consultingpartners.ifatccompanion.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.h3consultingpartners.ifatccompanion.ui.theme.IFATCTheme

/**
 * The shared building blocks of the app's screens, ported from
 * `IFATCCompanion/Views/Components.swift`.
 *
 * The iOS versions are SwiftUI views over a dark card aesthetic. These keep the same
 * shapes, weights and roles while being built from Material 3 surfaces, so they look
 * native on Android without losing the app's visual identity.
 *
 * Everything here is a pure composable — state in, callbacks out, no Android
 * dependencies — so `settings-uicheck.gradle.kts` can type-check it without the SDK.
 */

/** Traffic-light status used across the dashboard. */
enum class StatusLevel {
    GREEN,
    AMBER,
    RED,
    NEUTRAL,
    ;

    @Composable
    fun color(): Color = when (this) {
        GREEN -> IFATCTheme.semantic.connected
        AMBER -> IFATCTheme.semantic.connecting
        RED -> IFATCTheme.semantic.disconnected
        NEUTRAL -> MaterialTheme.colorScheme.outline
    }
}

/** A rounded card container. */
@Composable
fun Card(
    modifier: Modifier = Modifier,
    title: String? = null,
    icon: ImageVector? = null,
    content: @Composable ColumnScope.() -> Unit,
) {
    Surface(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(CARD_CORNER_RADIUS),
        color = MaterialTheme.colorScheme.surfaceContainer,
    ) {
        Column(
            modifier = Modifier.padding(CARD_PADDING),
            verticalArrangement = Arrangement.spacedBy(CARD_SPACING),
        ) {
            if (title != null) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    if (icon != null) {
                        Icon(
                            imageVector = icon,
                            contentDescription = null,
                            modifier = Modifier.size(18.dp),
                            tint = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    Text(
                        text = title,
                        style = MaterialTheme.typography.titleSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
            content()
        }
    }
}

/** A status chip with a coloured dot. */
@Composable
fun StatusPill(
    text: String,
    level: StatusLevel,
    modifier: Modifier = Modifier,
    icon: ImageVector? = null,
) {
    val tint = level.color()
    Surface(
        modifier = modifier,
        shape = CircleShape,
        color = tint.copy(alpha = PILL_FILL_ALPHA),
        border = BorderStroke(1.dp, tint.copy(alpha = PILL_STROKE_ALPHA)),
    ) {
        Row(
            modifier = Modifier.padding(horizontal = 12.dp, vertical = 7.dp),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Box(
                modifier = Modifier
                    .size(10.dp)
                    .background(tint, CircleShape),
            )
            if (icon != null) {
                Icon(
                    imageVector = icon,
                    contentDescription = null,
                    modifier = Modifier.size(14.dp),
                )
            }
            Text(
                text = text,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.SemiBold,
            )
        }
    }
}

/** A label/value row for data panels. */
@Composable
fun DataRow(
    label: String,
    value: String,
    modifier: Modifier = Modifier,
    icon: ImageVector? = null,
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(6.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (icon != null) {
                Icon(
                    imageVector = icon,
                    contentDescription = null,
                    modifier = Modifier.size(18.dp),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Spacer(Modifier.width(4.dp))
            }
            Text(
                text = label,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        Spacer(Modifier.weight(1f))
        Text(
            text = value,
            style = MaterialTheme.typography.bodyMedium,
            fontWeight = FontWeight.SemiBold,
            textAlign = TextAlign.End,
        )
    }
}

/**
 * Large action button suitable for one-handed use while flying.
 *
 * The 64 dp minimum height is carried over from iOS deliberately: it is comfortably
 * above Android's 48 dp touch-target minimum, which matters when the pilot is tapping it
 * on a phone propped next to a tablet mid-flight.
 */
@Composable
fun ActionButton(
    title: String,
    icon: ImageVector,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    tint: Color = MaterialTheme.colorScheme.primary,
    enabled: Boolean = true,
) {
    OutlinedButton(
        onClick = onClick,
        modifier = modifier.heightIn(min = ACTION_BUTTON_MIN_HEIGHT),
        enabled = enabled,
        shape = RoundedCornerShape(12.dp),
        border = BorderStroke(1.dp, tint.copy(alpha = 0.5f)),
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                modifier = Modifier.size(22.dp),
                tint = if (enabled) tint else LocalContentColor.current,
            )
            Text(
                text = title,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.SemiBold,
                textAlign = TextAlign.Center,
                maxLines = 2,
                overflow = TextOverflow.Ellipsis,
            )
        }
    }
}

/**
 * A frequency-tune button: facility name plus the frequency it's reached on,
 * highlighted while it's the controller currently being worked. Dimmed once the facility
 * has no further call in the flight.
 */
@Composable
fun FrequencyButton(
    title: String,
    icon: ImageVector,
    frequency: String,
    active: Boolean,
    enabled: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    // One button, styled two ways — filled while it is the controller being worked,
    // outlined otherwise. It stays a single Button rather than a Surface wrapping one so
    // TalkBack announces it as a button exactly once, and the whole label (facility and
    // frequency) is read together.
    val label = "$title, $frequency"
    val colors = if (active) {
        ButtonDefaults.buttonColors()
    } else {
        ButtonDefaults.outlinedButtonColors()
    }
    Button(
        onClick = onClick,
        modifier = modifier
            .heightIn(min = FREQUENCY_BUTTON_MIN_HEIGHT)
            .semantics { contentDescription = label },
        enabled = enabled,
        shape = RoundedCornerShape(12.dp),
        colors = colors,
        border = if (active) {
            null
        } else {
            BorderStroke(1.dp, MaterialTheme.colorScheme.primary.copy(alpha = 0.5f))
        },
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                modifier = Modifier.size(22.dp),
            )
            Text(
                text = title,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.SemiBold,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(
                text = frequency,
                style = MaterialTheme.typography.labelSmall,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }
    }
}

private val CARD_CORNER_RADIUS = 16.dp
private val CARD_PADDING = 16.dp
private val CARD_SPACING = 10.dp
private val ACTION_BUTTON_MIN_HEIGHT = 64.dp
private val FREQUENCY_BUTTON_MIN_HEIGHT = 72.dp
private const val PILL_FILL_ALPHA = 0.18f
private const val PILL_STROKE_ALPHA = 0.5f
